core/store/src/trie/trie_recording.rs

```diff
@@ -1,31 +1,37 @@
+use super::{Trie, TrieChanges, TrieRefcountDeltaMap};
 use crate::{NibbleSlice, PartialStorage, RawTrieNode, RawTrieNodeWithSize};
 use borsh::BorshDeserialize;
 use near_primitives::hash::CryptoHash;
 use near_primitives::state::PartialState;
 use near_primitives::trie_key::col::ALL_COLUMNS_WITH_NAMES;
 use near_primitives::types::AccountId;
-use std::collections::{HashMap, HashSet, VecDeque};
+use std::collections::{HashSet, VecDeque};
 use std::sync::Arc;
-
-use super::{Trie, TrieChanges, TrieRefcountDeltaMap};
+use std::sync::atomic::{AtomicUsize, Ordering};
 
 /// A simple struct to capture a state proof as it's being accumulated.
 pub struct TrieRecorder {
-    recorded: HashMap<CryptoHash, TrieNodeWithRefcount>,
-    size: usize,
-    /// Size of the recorded state proof plus some additional size added to cover removals and contract code.
-    /// An upper-bound estimation of the true recorded size after finalization.
-    /// See https://github.com/near/nearcore/issues/10890 and https://github.com/near/nearcore/pull/11000 for details.
-    upper_bound_size: usize,
-    /// Soft limit on the maximum size of the state proof that can be recorded.
-    proof_size_limit: Option<usize>,
+    recorded: dashmap::DashMap<CryptoHash, TrieNodeWithRefcount>,
+    size: AtomicUsize,
+    /// Size of the recorded state proof plus some additional size added to cover removals and
+    /// contract code.
+    ///
+    /// An upper-bound estimation of the true recorded size after finalization. See
+    /// https://github.com/near/nearcore/issues/10890 and
+    /// https://github.com/near/nearcore/pull/11000 for details.
+    upper_bound_size: AtomicUsize,
+    /// Limit on the maximum size of the state proof that can be recorded.
+    ///
+    /// This may get set to u64::MAX to effectively impose no useful limit.
+    proof_size_limit: u64,
```

Comment:
I figured that u64::MAX is high enough a limit to effectively mean a `None` which simplifies code somewhat IMO.